ament_cmake_auto/cmake/ament_auto_add_executable.cmake

```diff
@@ -73,11 +73,15 @@ macro(ament_auto_add_executable target)
   # link against other libraries of this package
   if(NOT ${PROJECT_NAME}_LIBRARIES STREQUAL "" AND
       NOT ARG_NO_TARGET_LINK_LIBRARIES)
-    target_link_libraries("${target}" ${${PROJECT_NAME}_LIBRARIES})
+    foreach(lib ${${PROJECT_NAME}_LIBRARIES})
+      get_target_property(lib_include_dirs ${lib} INTERFACE_INCLUDE_DIRECTORIES)
+      target_include_directories("${target}" SYSTEM PRIVATE ${lib_include_dirs})
```

Comment:
Use of `SYSTEM` when linking targets that are within the current package feels odd to me.
I'm also somewhat concerned by the fact that `SYSTEM` affects the search order:

> If SYSTEM is specified, the compiler will be told the directories are meant as system include directories on some platforms. This may have effects such as suppressing warnings or skipping the contained headers in dependency calculations (see compiler documentation). Additionally, system include directories are searched after normal include directories regardless of the order specified.

Reply:
@nuclearsandwich The majority of the issues that I've encountered were when building Autoware.Auto in Galactic or (in an unsupported environment) building Foxy-released packages in an Ubuntu 18.04-based environment. All of the issues that this fixed were related to linked packages causing errors. As an example, if a package links against PCL, IMO, PCL should not be searched for warnings or errors. Adding the SYSTEM keyword avoids this completely and I have not found another way to do so.

Regarding targets in the current package, you are correct - I don't believe these should be linked as SYSTEM either. Let me test this with just the modification to `ament_target_dependencies` and get back to you.

Reply:
> I haven't reviewed the PR - just a quick comment that Imported targets (targets from other packages) have their include directories added as SYSTEM by default, while non-imported targets don't.
> 
> https://cmake.org/cmake/help/v3.0/prop_tgt/NO_SYSTEM_FROM_IMPORTED.html

Tangentially, CMake was not properly using the external/system includes feature from MSVC properly until a few hours ago, so @JWhitleyWork may very well have been facing the "targets-apparently-not-imported-as-system" problem on Windows.

https://gitlab.kitware.com/cmake/cmake/-/issues/17904

Reply:
@aprotyas This was not on Windows so it is unrelated.

Reply:
@nuclearsandwich It turns out that this change was not necessary. I have reverted it. The remaining change is still necessary to fix the issue.